File: pawsec-extension/src/analyzers/business_data_detector.js
/**
 * PawSec Business Data Detector (browser port of business_data_detector.py)
 *
 * Detects business entity identifiers: EIN (US), DUNS, EU VAT (27 countries),
 * UK Company Number, ABN/ACN (Australia), Canada BN.
 *
 * Scope: Organisation/entity-level identifiers only.
 * Individual identity documents (driver's licences, passports, etc.) are handled
 * by pii_detector.js.
 *
 * Returns:
 *   detected    — true if any identifiers found
 *   findings    — array of { type, name, jurisdiction, severity, value_masked, position }
 *   count       — total findings
 *   risk_level  — SAFE / MEDIUM / HIGH / CRITICAL
 */

// ── Helper: mask sensitive values ─────────────────────────────────────────
function _maskValue(value) {
  if (value.length <= 4) return '*'.repeat(value.length);
  return value.slice(0, 2) + '*'.repeat(value.length - 4) + value.slice(-2);
}

// ── Business identifier patterns ──────────────────────────────────────────
const PATTERNS = [
  // US EIN (labeled)
  {
    type: 'ein',
    name: 'US Employer Identification Number (EIN)',
    jurisdiction: 'United States',
    severity: 'HIGH',
    re: /(?:EIN|Employer\s+Identification\s+(?:Number|No\.?)|Federal\s+Tax\s+ID|FEIN)\s*[:\-#]?\s*(\b\d{2}-\d{7}\b)/gi,
  },
  // DUNS Number
  {
    type: 'duns',
    name: 'DUNS Number (Dun & Bradstreet)',
    jurisdiction: 'International',
    severity: 'MEDIUM',
    re: /(?:DUNS|D[-&]U[-&]N[-&]S|Dun\s+&\s+Bradstreet)\s*(?:Number|No\.?|#)?\s*[:\-]?\s*(\b\d{2}-\d{3}-\d{4}\b|\b\d{9}\b)/gi,
  },
  // EU VAT (27 member states)
  {
    type: 'vat_eu',
    name: 'EU VAT Number',
    jurisdiction: 'European Union',
    severity: 'MEDIUM',
    re: /(?:VAT|Value\s+Added\s+Tax)\s*(?:Number|No\.?|#|ID|Registration)?\s*[:\-]?\s*(AT\s*U\d{8}|BE\s*0\d{9}|BG\s*\d{9,10}|CY\s*\d{8}[A-Z]|CZ\s*\d{8,10}|DE\s*\d{9}|DK\s*\d{8}|EE\s*\d{9}|EL\s*\d{9}|ES\s*[A-Z0-9]\d{7}[A-Z0-9]|FI\s*\d{8}|FR\s*[A-HJ-NP-Z0-9]{2}\d{9}|HR\s*\d{11}|HU\s*\d{8}|IE\s*\d{7}[A-Z]{1,2}|IT\s*\d{11}|LT\s*(?:\d{9}|\d{12})|LU\s*\d{8}|LV\s*\d{11}|MT\s*\d{8}|NL\s*\d{9}B\d{2}|PL\s*\d{10}|PT\s*\d{9}|RO\s*\d{2,10}|SE\s*\d{12}|SI\s*\d{8}|SK\s*\d{10})/gi,
  },
  // UK Company Number
  {
    type: 'uk_company',
    name: 'UK Company Registration Number',
    jurisdiction: 'United Kingdom',
    severity: 'MEDIUM',
    re: /(?:Company\s+(?:Number|No\.?|Registration)|Companies\s+House|CRN)\s*[:\-]?\s*(\b(?:[A-Z]{2}|[A-Z])?\d{6,8}\b)/gi,
  },
  // Australian Business Number
  {
    type: 'abn',
    name: 'Australian Business Number (ABN)',
    jurisdiction: 'Australia',
    severity: 'MEDIUM',
    re: /(?:ABN|Australian\s+Business\s+Number)\s*[:\-]?\s*(\b\d{2}\s?\d{3}\s?\d{3}\s?\d{3}\b)/gi,
  },
  // Australian Company Number
  {
    type: 'acn',
    name: 'Australian Company Number (ACN)',
    jurisdiction: 'Australia',
    severity: 'MEDIUM',
    re: /(?:ACN|Australian\s+Company\s+Number)\s*[:\-]?\s*(\b\d{3}\s?\d{3}\s?\d{3}\b)/gi,
  },
  // Canada Business Number
  {
    type: 'canada_bn',
    name: 'Canada Business Number (BN)',
    jurisdiction: 'Canada',
    severity: 'MEDIUM',
    re: /(?:BN|Business\s+Number|CRA\s+Number)\s*[:\-]?\s*(\b\d{9}\s?(?:RT|RP|RC|RZ)\s?\d{4}\b)/gi,
  },
];

// EIN standalone context (labeled separately in Python)
const EIN_CONTEXT_RE  = /(?:EIN|FEIN|Tax\s+ID|Employer\s+ID)/gi;
const EIN_STANDALONE  = /\b(\d{2}-\d{7})\b/g;

/**
 * @param {string} text
 * @returns {{
 *   detected: boolean,
 *   findings: Array,
 *   count: number,
 *   risk_level: string
 * }}
 */
export function analyzeBusinessData(text) {
  const findings = [];
  const seenPositions = [];

  for (const spec of PATTERNS) {
    spec.re.lastIndex = 0;
    for (const m of text.matchAll(spec.re)) {
      const pos = m.index;
      if (seenPositions.some(p => Math.abs(pos - p) < 10)) continue;
      seenPositions.push(pos);

      const value = m[1] ?? m[0];
      findings.push({
        type:         spec.type,
        name:         spec.name,
        jurisdiction: spec.jurisdiction,
        severity:     spec.severity,
        value_masked: _maskValue(value),
        position:     pos,
      });
    }

    // EIN standalone context check (mirrors Python standalone logic)
    if (spec.type === 'ein') {
      EIN_CONTEXT_RE.lastIndex = 0;
      for (const ctxM of text.matchAll(EIN_CONTEXT_RE)) {
        const windowStart = Math.max(0, ctxM.index - 5);
        const windowEnd   = Math.min(text.length, ctxM.index + ctxM[0].length + 30);
        const window      = text.slice(windowStart, windowEnd);
        EIN_STANDALONE.lastIndex = 0;
        for (const einM of window.matchAll(EIN_STANDALONE)) {
          const absPos = windowStart + einM.index;
          if (seenPositions.some(p => Math.abs(absPos - p) < 10)) continue;
          seenPositions.push(absPos);
          findings.push({
            type:         'ein',
            name:         'US Employer Identification Number (EIN)',
            jurisdiction: 'United States',
            severity:     'HIGH',
            value_masked: _maskValue(einM[1]),
            position:     absPos,
          });
        }
      }
    }
  }

  const risk_level = findings.length
    ? (findings.some(f => f.severity === 'CRITICAL') ? 'CRITICAL'
       : findings.some(f => f.severity === 'HIGH')   ? 'HIGH'
       : 'MEDIUM')
    : 'SAFE';

  return {
    detected:   findings.length > 0,
    findings,
    count:      findings.length,
    risk_level,
  };
}

File: skills/business-data-detector/tools/business_data_detector.py
"""
Business Data Detector
Detects business entity identifiers: EIN (US), DUNS numbers, VAT numbers (EU),
ABN/ACN (Australia), UK company numbers, and Canada Business Numbers.

Scope: Organisation/entity-level identifiers only.
Individual identity documents (driver's licences, passports, etc.) are handled
by the pii-detector skill.
"""

import re
import json
from datetime import datetime
from typing import Dict, List


class BusinessDataDetector:
    """Detects business entity identifiers (EIN, DUNS, VAT, company numbers)."""

    def __init__(self):
        self._patterns = self._load_patterns()

    def _load_patterns(self) -> List[Dict]:
        return [
            # ── US Employer Identification Number (EIN) ───────────────────────────
            {
                "type": "ein",
                "name": "US Employer Identification Number (EIN)",
                "jurisdiction": "United States",
                "severity": "HIGH",
                "pattern": re.compile(
                    r'(?:EIN|Employer\s+Identification\s+(?:Number|No\.?)|'
                    r'Federal\s+Tax\s+ID|FEIN)\s*[:\-#]?\s*'
                    r'(\b\d{2}-\d{7}\b)',
                    re.IGNORECASE
                ),
                "standalone": re.compile(r'\b(\d{2}-\d{7})\b')
            },
            # ── DUNS Number ──────────────────────────────────────────────────────
            {
                "type": "duns",
                "name": "DUNS Number (Dun & Bradstreet)",
                "jurisdiction": "International",
                "severity": "MEDIUM",
                "pattern": re.compile(
                    r'(?:DUNS|D[-&]U[-&]N[-&]S|Dun\s+&\s+Bradstreet)\s*(?:Number|No\.?|#)?\s*[:\-]?\s*'
                    r'(\b\d{2}-\d{3}-\d{4}\b|\b\d{9}\b)',
                    re.IGNORECASE
                )
            },
            # ── VAT Numbers (EU) ─────────────────────────────────────────────────
            {
                "type": "vat_eu",
                "name": "EU VAT Number",
                "jurisdiction": "European Union",
                "severity": "MEDIUM",
                "pattern": re.compile(
                    r'(?:VAT|Value\s+Added\s+Tax)\s*(?:Number|No\.?|#|ID|Registration)?\s*[:\-]?\s*'
                    r'('
                    r'AT\s*U\d{8}|'            # Austria
                    r'BE\s*0\d{9}|'            # Belgium
                    r'BG\s*\d{9,10}|'          # Bulgaria
                    r'CY\s*\d{8}[A-Z]|'        # Cyprus
                    r'CZ\s*\d{8,10}|'          # Czech Republic
                    r'DE\s*\d{9}|'             # Germany
                    r'DK\s*\d{8}|'             # Denmark
                    r'EE\s*\d{9}|'             # Estonia
                    r'EL\s*\d{9}|'             # Greece
                    r'ES\s*[A-Z0-9]\d{7}[A-Z0-9]|'  # Spain
                    r'FI\s*\d{8}|'             # Finland
                    r'FR\s*[A-HJ-NP-Z0-9]{2}\d{9}|'  # France
                    r'HR\s*\d{11}|'            # Croatia
                    r'HU\s*\d{8}|'             # Hungary
                    r'IE\s*\d{7}[A-Z]{1,2}|'  # Ireland
                    r'IT\s*\d{11}|'            # Italy
                    r'LT\s*(?:\d{9}|\d{12})|' # Lithuania
                    r'LU\s*\d{8}|'             # Luxembourg
                    r'LV\s*\d{11}|'            # Latvia
                    r'MT\s*\d{8}|'             # Malta
                    r'NL\s*\d{9}B\d{2}|'       # Netherlands
                    r'PL\s*\d{10}|'            # Poland
                    r'PT\s*\d{9}|'             # Portugal
                    r'RO\s*\d{2,10}|'          # Romania
                    r'SE\s*\d{12}|'            # Sweden
                    r'SI\s*\d{8}|'             # Slovenia
                    r'SK\s*\d{10}'             # Slovakia
                    r')',
                    re.IGNORECASE
                )
            },
            # ── UK Company Number ─────────────────────────────────────────────────
            {
                "type": "uk_company",
                "name": "UK Company Registration Number",
                "jurisdiction": "United Kingdom",
                "severity": "MEDIUM",
                "pattern": re.compile(
                    r'(?:Company\s+(?:Number|No\.?|Registration)|Companies\s+House|CRN)\s*[:\-]?\s*'
                    r'(\b(?:[A-Z]{2}|[A-Z])?\d{6,8}\b)',
                    re.IGNORECASE
                )
            },
            # ── Australian Business Number (ABN) ──────────────────────────────────
            {
                "type": "abn",
                "name": "Australian Business Number (ABN)",
                "jurisdiction": "Australia",
                "severity": "MEDIUM",
                "pattern": re.compile(
                    r'(?:ABN|Australian\s+Business\s+Number)\s*[:\-]?\s*'
                    r'(\b\d{2}\s?\d{3}\s?\d{3}\s?\d{3}\b)',
                    re.IGNORECASE
                )
            },
            # ── Australian Company Number (ACN) ───────────────────────────────────
            {
                "type": "acn",
                "name": "Australian Company Number (ACN)",
                "jurisdiction": "Australia",
                "severity": "MEDIUM",
                "pattern": re.compile(
                    r'(?:ACN|Australian\s+Company\s+Number)\s*[:\-]?\s*'
                    r'(\b\d{3}\s?\d{3}\s?\d{3}\b)',
                    re.IGNORECASE
                )
            },
            # ── Canadian Business Number ──────────────────────────────────────────
            {
                "type": "canada_bn",
                "name": "Canada Business Number (BN)",
                "jurisdiction": "Canada",
                "severity": "MEDIUM",
                "pattern": re.compile(
                    r'(?:BN|Business\s+Number|CRA\s+Number)\s*[:\-]?\s*'
                    r'(\b\d{9}\s?(?:RT|RP|RC|RZ)\s?\d{4}\b)',
                    re.IGNORECASE
                )
            },
        ]

    @staticmethod
    def _mask_value(value: str) -> str:
        if len(value) <= 4:
            return '*' * len(value)
        return value[:2] + '*' * (len(value) - 4) + value[-2:]

    def analyze(self, text: str) -> Dict:
        """Analyze text for business entity identifiers."""
        findings = []
        seen_positions = []

        for spec in self._patterns:
            for match in spec["pattern"].finditer(text):
                pos = match.start()
                if any(abs(pos - p) < 10 for p in seen_positions):
                    continue
                seen_positions.append(pos)

                value = match.group(1) if match.lastindex else match.group(0)
                findings.append({
                    "type": spec["type"],
                    "name": spec["name"],
                    "jurisdiction": spec["jurisdiction"],
                    "severity": spec["severity"],
                    "value_masked": self._mask_value(value),
                    "position": pos
                })

            # Also check standalone EIN pattern with context
            if spec["type"] == "ein" and "standalone" in spec:
                ein_context = re.compile(
                    r'(?:EIN|FEIN|Tax\s+ID|Employer\s+ID)',
                    re.IGNORECASE
                )
                for ctx_m in ein_context.finditer(text):
                    window_start = max(0, ctx_m.start() - 5)
                    window_end = min(len(text), ctx_m.end() + 30)
                    window = text[window_start:window_end]
                    for ein_m in spec["standalone"].finditer(window):
                        abs_pos = window_start + ein_m.start()
                        if any(abs(abs_pos - p) < 10 for p in seen_positions):
                            continue
                        seen_positions.append(abs_pos)
                        findings.append({
                            "type": "ein",
                            "name": "US Employer Identification Number (EIN)",
                            "jurisdiction": "United States",
                            "severity": "HIGH",
                            "value_masked": self._mask_value(ein_m.group(1)),
                            "position": abs_pos
                        })

        # Risk level
        if findings:
            severities = [f["severity"] for f in findings]
            if "CRITICAL" in severities:
                risk_level = "CRITICAL"
            elif "HIGH" in severities:
                risk_level = "HIGH"
            else:
                risk_level = "MEDIUM"
        else:
            risk_level = "SAFE"

        return {
            "detected": len(findings) > 0,
            "findings": findings,
            "count": len(findings),
            "risk_level": risk_level,
            "timestamp": datetime.utcnow().isoformat() + "Z"
        }

    def analyze_batch(self, texts: List[str]) -> List[Dict]:
        return [self.analyze(t) for t in texts]


# ── CLI entry point ──────────────────────────────────────────────────────────
if __name__ == "__main__":
    import sys

    detector = BusinessDataDetector()

    if len(sys.argv) > 1:
        text = " ".join(sys.argv[1:])
    else:
        print("Enter text to analyze (Ctrl+D/Ctrl+Z to end):")
        try:
            text = sys.stdin.read()
        except KeyboardInterrupt:
            sys.exit(0)

    result = detector.analyze(text)
    print(json.dumps(result, indent=2))
